### PR TITLE
Standardize RL unit tests

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tests/unit_tests/test_rl_utils.py [hunk 1]
```diff
@@ -66,281 +65,217 @@ def detokenize(self, tokens):
...
+    """
+    tp, pp = request.param
+    world_size = Utils.world_size
 
-    # Check that inference logprobs are being returned.
-    torch.testing.assert_close(inference_logprobs[0], torch.tensor([0.1, 0.2, 0.3]))
-    torch.testing.assert_close(inference_logprobs[1], torch.tensor([0.1, 0.2, 0.3]))
+    # Skip if world size is smaller than the parallelization requirements
+    required_size = tp * pp
+    if required_size > 0 and world_size < required_size:
```
Comment (yobibyte) [inline tests/unit_tests/test_rl_utils.py]: This seems dangerous to me, IMO, we should fail in this situation and either fix the failing test that breaks our assumptions on availability of GPUs, or flag this with the CI team that should give us compute.

**Comment by yobibyte** [comment]
Hey @tdene, amazing job, this looks so much cleaner now.

However, when I run your tests on an interactive node with 8 GPUs with 'just unit', this times out at initialization. Are we supposed to run that with torchrun now? Is this what CI expects?

File: tests/unit_tests/test_rl_utils.py (hunk 1 above)
Comment (tdene) [reply tests/unit_tests/test_rl_utils.py]: Resolved!

File: tests/unit_tests/rl/test_rl_utils.py [hunk 2]
```diff
@@ -66,281 +64,206 @@ def detokenize(self, tokens):
...
-        reward=0.14,
-        generation_mask=[False, True, True],
-        logprobs=[0.1, 0.2, 0.3],
-        env_id='MEGAENV',
-        problem_id="2",
-    )
-    rollouts = [[r1, r2]]
-    seq_len = 7
+@pytest.fixture
+def initialize_model_parallel(request):
```
Comment (jon-barker) [inline tests/unit_tests/rl/test_rl_utils.py]: It seems like this should live somewhere more general than this rl specific file. There's an existing Utils.initialize_model_parallel which you use here - can that not be extended to provide this functionality for all users?

File: tests/unit_tests/rl/test_rl_utils.py [hunk 3]
```diff
@@ -66,281 +64,206 @@ def detokenize(self, tokens):
...
-    global_vars.unset_global_variables()
-    global_vars.set_global_variables(args, build_tokenizer=False)
+        args = validate_args(args)
+        set_global_variables(args, False)
+        return args
 
-    model = MockModel()
-    tokenizer = MockTokenizer()
+    @pytest.mark.parametrize(
+        "initialize_model_parallel",
```
Comment (jon-barker) [inline tests/unit_tests/rl/test_rl_utils.py:121]: Why do we only test pp=1? if that's always going to be the case then this code can be simplified

File: tests/unit_tests/rl/test_rl_utils.py (hunk 2 above)
Comment (tdene) [reply tests/unit_tests/rl/test_rl_utils.py]: Agreed. Currently, each unit test independently redefines this method; it's done a couple of dozen times independently.

This kind of improvement should go into a comprehensive cleaning pass of all unit tests.

File: tests/unit_tests/rl/test_rl_utils.py (hunk 3 above)
Comment (tdene) [reply tests/unit_tests/rl/test_rl_utils.py:121]: Fixed! I had just not implemented the code to allow for PP>1, and that's done now.

Comment (jon-barker) [reply tests/unit_tests/rl/test_rl_utils.py:121]: thanks

File: tests/unit_tests/rl/test_rl_utils.py (hunk 2 above)
Comment (jon-barker) [reply tests/unit_tests/rl/test_rl_utils.py]: fair enough
